File: modal-js/src/queue.ts
// Queue object, to be used with Modal Queues.

import {
  ObjectCreationType,
  QueueNextItemsRequest,
} from "../proto/modal_proto/api";
import { getDefaultClient, type ModalClient } from "./client";
import { InvalidError, QueueEmptyError, QueueFullError } from "./errors";
import { dumps, loads } from "./pickle";
import { ClientError, Status } from "nice-grpc";
import { EphemeralHeartbeatManager } from "./ephemeral";

const queueInitialPutBackoff = 100; // 100 milliseconds
const queueDefaultPartitionTtl = 24 * 3600 * 1000; // 24 hours

/** Options for `client.queues.fromName()`. */
export type QueueFromNameOptions = {
  environment?: string;
  createIfMissing?: boolean;
};

/** Options for `client.queues.delete()`. */
export type QueueDeleteOptions = {
  environment?: string;
};

/** Options for `client.queues.ephemeral()`. */
export type QueueEphemeralOptions = {
  environment?: string;
};

/**
 * Service for managing Queues.
 */
export class QueueService {
  readonly #client: ModalClient;
  constructor(client: ModalClient) {
    this.#client = client;
  }

  /**
   * Create a nameless, temporary Queue.
   * You will need to call `closeEphemeral()` to delete the Queue.
   */
  async ephemeral(options: QueueEphemeralOptions = {}): Promise<Queue> {
    const resp = await this.#client.cpClient.queueGetOrCreate({
      objectCreationType: ObjectCreationType.OBJECT_CREATION_TYPE_EPHEMERAL,
      environmentName: this.#client.environmentName(options.environment),
    });

    const ephemeralHbManager = new EphemeralHeartbeatManager(() =>
      this.#client.cpClient.queueHeartbeat({ queueId: resp.queueId }),
    );

    return new Queue(this.#client, resp.queueId, undefined, ephemeralHbManager);
  }

  /**
   * Lookup a Queue by name.
   */
  async fromName(
    name: string,
    options: QueueFromNameOptions = {},
  ): Promise<Queue> {
    const resp = await this.#client.cpClient.queueGetOrCreate({
      deploymentName: name,
      objectCreationType: options.createIfMissing
        ? ObjectCreationType.OBJECT_CREATION_TYPE_CREATE_IF_MISSING
        : undefined,
      environmentName: this.#client.environmentName(options.environment),
    });
    return new Queue(this.#client, resp.queueId, name);
  }

  /**
   * Delete a Queue by name.
   */
  async delete(name: string, options: QueueDeleteOptions = {}): Promise<void> {
    const queue = await this.fromName(name, options);
    await this.#client.cpClient.queueDelete({ queueId: queue.queueId });
  }
}

/** Options to configure a `Queue.clear()` operation. */
export type QueueClearOptions = {
  /** Partition to clear, uses default partition if not set. */
  partition?: string;

  /** Set to clear all Queue partitions. */
  all?: boolean;
};

/** Options to configure a `Queue.get()` or `Queue.getMany()` operation. */
export type QueueGetOptions = {
  /** How long to wait if the Queue is empty (default: indefinite). */
  timeout?: number;

  /** Partition to fetch values from, uses default partition if not set. */
  partition?: string;
};

/** Options to configure a `Queue.put()` or `Queue.putMany()` operation. */
export type QueuePutOptions = {
  /** How long to wait if the Queue is full (default: indefinite). */
  timeout?: number;

  /** Partition to add items to, uses default partition if not set. */
  partition?: string;

  /** TTL for the partition in seconds (default: 1 day). */
  partitionTtl?: number;
};

/** Options to configure a `Queue.len()` operation. */
export type QueueLenOptions = {
  /** Partition to compute length, uses default partition if not set. */
  partition?: string;

  /** Return the total length across all partitions. */
  total?: boolean;
};

/** Options to configure a `Queue.iterate()` operation. */
export type QueueIterateOptions = {
  /** How long to wait between successive items before exiting iteration (default: 0). */
  itemPollTimeout?: number;

  /** Partition to iterate, uses default partition if not set. */
  partition?: string;
};

/**
 * Distributed, FIFO queue for data flow in Modal Apps.
 */
export class Queue {
  readonly #client: ModalClient;
  readonly queueId: string;
  readonly name?: string;
  readonly #ephemeralHbManager?: EphemeralHeartbeatManager;

  /** @ignore */
  constructor(
    client: ModalClient,
    queueId: string,
    name?: string,
    ephemeralHbManager?: EphemeralHeartbeatManager,
  ) {
    this.#client = client;
    this.queueId = queueId;
    this.name = name;
    this.#ephemeralHbManager = ephemeralHbManager;
  }

  static #validatePartitionKey(partition: string | undefined): Uint8Array {
    if (partition) {
      const partitionKey = new TextEncoder().encode(partition);
      if (partitionKey.length === 0 || partitionKey.length > 64) {
        throw new InvalidError(
          "Queue partition key must be between 1 and 64 bytes.",
        );
      }
      return partitionKey;
    }
    return new Uint8Array();
  }

  /**
   * @deprecated Use `client.queues.ephemeral()` instead.
   */
  static async ephemeral(options: QueueEphemeralOptions = {}): Promise<Queue> {
    return getDefaultClient().queues.ephemeral(options);
  }

  /** Delete the ephemeral Queue. Only usable with ephemeral Queues. */
  closeEphemeral(): void {
    if (this.#ephemeralHbManager) {
      this.#ephemeralHbManager.stop();
    } else {
      throw new InvalidError("Queue is not ephemeral.");
    }
  }

  /**
   * @deprecated Use `client.queues.fromName()` instead.
   */
  static async lookup(
    name: string,
    options: QueueFromNameOptions = {},
  ): Promise<Queue> {
    return getDefaultClient().queues.fromName(name, options);
  }

  /**
   * @deprecated Use `client.queues.delete()` instead.
   */
  static async delete(
    name: string,
    options: QueueDeleteOptions = {},
  ): Promise<void> {
    return getDefaultClient().queues.delete(name, options);
  }

  /**
   * Remove all objects from a Queue partition.
   */
  async clear(options: QueueClearOptions = {}): Promise<void> {
    if (options.partition && options.all) {
      throw new InvalidError(
        "Partition must be null when requesting to clear all.",
      );
    }
    await this.#client.cpClient.queueClear({
      queueId: this.queueId,
      partitionKey: Queue.#validatePartitionKey(options.partition),
      allPartitions: options.all,
    });
  }

  async #get(n: number, partition?: string, timeout?: number): Promise<any[]> {
    const partitionKey = Queue.#validatePartitionKey(partition);

    const startTime = Date.now();
    let pollTimeout = 50_000;
    if (timeout !== undefined) {
      pollTimeout = Math.min(pollTimeout, timeout);
    }

    while (true) {
      const response = await this.#client.cpClient.queueGet({
        queueId: this.queueId,
        partitionKey,
        timeout: pollTimeout / 1000,
        nValues: n,
      });
      if (response.values && response.values.length > 0) {
        return response.values.map((value) => loads(value));
      }
      if (timeout !== undefined) {
        const remaining = timeout - (Date.now() - startTime);
        if (remaining <= 0) {
          const message = `Queue ${this.queueId} did not return values within ${timeout}ms.`;
          throw new QueueEmptyError(message);
        }
        pollTimeout = Math.min(pollTimeout, remaining);
      }
    }
  }

  /**
   * Remove and return the next object from the Queue.
   *
   * By default, this will wait until at least one item is present in the Queue.
   * If `timeout` is set, raises `QueueEmptyError` if no items are available
   * within that timeout in milliseconds.
   */
  async get(options: QueueGetOptions = {}): Promise<any | null> {
    const values = await this.#get(1, options.partition, options.timeout);
    return values[0]; // Must have length >= 1 if returned.
  }

  /**
   * Remove and return up to `n` objects from the Queue.
   *
   * By default, this will wait until at least one item is present in the Queue.
   * If `timeout` is set, raises `QueueEmptyError` if no items are available
   * within that timeout in milliseconds.
   */
  async getMany(n: number, options: QueueGetOptions = {}): Promise<any[]> {
    return await this.#get(n, options.partition, options.timeout);
  }

  async #put(
    values: any[],
    timeout?: number,
    partition?: string,
    partitionTtl?: number,
  ): Promise<void> {
    const valuesEncoded = values.map((v) => dumps(v));
    const partitionKey = Queue.#validatePartitionKey(partition);

    let delay = queueInitialPutBackoff;
    const deadline = timeout ? Date.now() + timeout : undefined;
    while (true) {
      try {
        await this.#client.cpClient.queuePut({
          queueId: this.queueId,
          values: valuesEncoded,
          partitionKey,
          partitionTtlSeconds:
            (partitionTtl || queueDefaultPartitionTtl) / 1000,
        });
        break;
      } catch (e) {
        if (e instanceof ClientError && e.code === Status.RESOURCE_EXHAUSTED) {
          // Queue is full, retry with exponential backoff up to the deadline.
          delay = Math.min(delay * 2, 30_000);
          if (deadline !== undefined) {
            const remaining = deadline - Date.now();
            if (remaining <= 0)
              throw new QueueFullError(`Put failed on ${this.queueId}.`);
            delay = Math.min(delay, remaining);
          }
          await new Promise((resolve) => setTimeout(resolve, delay));
        } else {
          throw e;
        }
      }
    }
  }

  /**
   * Add an item to the end of the Queue.
   *
   * If the Queue is full, this will retry with exponential backoff until the
   * provided `timeout` is reached, or indefinitely if `timeout` is not set.
   * Raises `QueueFullError` if the Queue is still full after the timeout.
   */
  async put(v: any, options: QueuePutOptions = {}): Promise<void> {
    await this.#put(
      [v],
      options.timeout,
      options.partition,
      options.partitionTtl,
    );
  }

  /**
   * Add several items to the end of the Queue.
   *
   * If the Queue is full, this will retry with exponential backoff until the
   * provided `timeout` is reached, or indefinitely if `timeout` is not set.
   * Raises `QueueFullError` if the Queue is still full after the timeout.
   */
  async putMany(values: any[], options: QueuePutOptions = {}): Promise<void> {
    await this.#put(
      values,
      options.timeout,
      options.partition,
      options.partitionTtl,
    );
  }

  /** Return the number of objects in the Queue. */
  async len(options: QueueLenOptions = {}): Promise<number> {
    if (options.partition && options.total) {
      throw new InvalidError(
        "Partition must be null when requesting total length.",
      );
    }
    const resp = await this.#client.cpClient.queueLen({
      queueId: this.queueId,
      partitionKey: Queue.#validatePartitionKey(options.partition),
      total: options.total,
    });
    return resp.len;
  }

  /** Iterate through items in a Queue without mutation. */
  async *iterate(
    options: QueueIterateOptions = {},
  ): AsyncGenerator<any, void, unknown> {
    const { partition, itemPollTimeout = 0 } = options;

    let lastEntryId = undefined;
    const validatedPartitionKey = Queue.#validatePartitionKey(partition);
    let fetchDeadline = Date.now() + itemPollTimeout;

    const maxPollDuration = 30_000;
    while (true) {
      const pollDuration = Math.max(
        0.0,
        Math.min(maxPollDuration, fetchDeadline - Date.now()),
      );
      const request: QueueNextItemsRequest = {
        queueId: this.queueId,
        partitionKey: validatedPartitionKey,
        itemPollTimeout: pollDuration / 1000,
        lastEntryId: lastEntryId || "",
      };

      const response = await this.#client.cpClient.queueNextItems(request);
      if (response.items && response.items.length > 0) {
        for (const item of response.items) {
          yield loads(item.value);
          lastEntryId = item.entryId;
        }
        fetchDeadline = Date.now() + itemPollTimeout;
      } else if (Date.now() > fetchDeadline) {
        break;
      }
    }
  }
}

File: modal-go/queue.go
package modal

// Queue object, to be used with Modal Queues.

import (
	"context"
	"fmt"
	"iter"
	"time"

	pb "github.com/modal-labs/libmodal/modal-go/proto/modal_proto"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

// QueueService provides Queue related operations.
type QueueService struct{ client *Client }

const queueInitialPutBackoff = 100 * time.Millisecond
const queueDefaultPartitionTtl = 24 * time.Hour

func validatePartitionKey(partition string) ([]byte, error) {
	if partition == "" {
		return nil, nil // default partition
	}
	b := []byte(partition)
	if len(b) == 0 || len(b) > 64 {
		return nil, InvalidError{"Queue partition key must be 1–64 bytes long"}
	}
	return b, nil
}

type QueueClearOptions struct {
	Partition string // partition to clear (default "")
	All       bool   // clear *all* partitions (mutually exclusive with Partition)
}

type QueueGetOptions struct {
	Timeout   *time.Duration // wait max (nil = indefinitely)
	Partition string
}

type QueuePutOptions struct {
	Timeout      *time.Duration // max wait for space (nil = indefinitely)
	Partition    string
	PartitionTtl time.Duration // ttl for the *partition* (default 24h)
}

type QueueLenOptions struct {
	Partition string
	Total     bool // total across all partitions (mutually exclusive with Partition)
}

type QueueIterateOptions struct {
	ItemPollTimeout time.Duration // exit if no new items within this period
	Partition       string
}

// Queue is a distributed, FIFO queue for data flow in Modal Apps.
type Queue struct {
	QueueId         string
	Name            string
	cancelEphemeral context.CancelFunc

	client *Client
}

// Ephemeral creates a nameless, temporary Queue, that persists until CloseEphemeral is called, or the process exits.
func (s *QueueService) Ephemeral(ctx context.Context, options *EphemeralOptions) (*Queue, error) {
	if options == nil {
		options = &EphemeralOptions{}
	}

	resp, err := s.client.cpClient.QueueGetOrCreate(ctx, pb.QueueGetOrCreateRequest_builder{
		ObjectCreationType: pb.ObjectCreationType_OBJECT_CREATION_TYPE_EPHEMERAL,
		EnvironmentName:    environmentName(options.Environment, s.client.profile),
	}.Build())
	if err != nil {
		return nil, err
	}

	ephemeralCtx, cancel := context.WithCancel(context.Background())
	startEphemeralHeartbeat(ephemeralCtx, func() error {
		_, err := s.client.cpClient.QueueHeartbeat(ephemeralCtx, pb.QueueHeartbeatRequest_builder{
			QueueId: resp.GetQueueId(),
		}.Build())
		return err
	})

	q := &Queue{
		QueueId:         resp.GetQueueId(),
		cancelEphemeral: cancel,
		client:          s.client,
	}

	return q, nil
}

// CloseEphemeral deletes an ephemeral Queue, only used with QueueEphemeral.
func (q *Queue) CloseEphemeral() {
	if q.cancelEphemeral != nil {
		q.cancelEphemeral()
	} else {
		// We panic in this case because of invalid usage. In general, methods
		// used with `defer` like CloseEphemeral should not return errors.
		panic(fmt.Sprintf("Queue %s is not ephemeral", q.QueueId))
	}
}

// Lookup returns a handle to a (possibly new) Queue by name.
func (s *QueueService) Lookup(ctx context.Context, name string, options *LookupOptions) (*Queue, error) {
	if options == nil {
		options = &LookupOptions{}
	}

	creationType := pb.ObjectCreationType_OBJECT_CREATION_TYPE_UNSPECIFIED
	if options.CreateIfMissing {
		creationType = pb.ObjectCreationType_OBJECT_CREATION_TYPE_CREATE_IF_MISSING
	}

	resp, err := s.client.cpClient.QueueGetOrCreate(ctx, pb.QueueGetOrCreateRequest_builder{
		DeploymentName:     name,
		EnvironmentName:    environmentName(options.Environment, s.client.profile),
		ObjectCreationType: creationType,
	}.Build())
	if err != nil {
		return nil, err
	}
	return &Queue{
		QueueId:         resp.GetQueueId(),
		Name:            name,
		cancelEphemeral: nil,
		client:          s.client,
	}, nil
}

// Delete removes a Queue by name.
func (s *QueueService) Delete(ctx context.Context, name string, options *DeleteOptions) error {
	if options == nil {
		options = &DeleteOptions{}
	}

	q, err := s.Lookup(ctx, name, &LookupOptions{Environment: options.Environment})
	if err != nil {
		return err
	}
	_, err = s.client.cpClient.QueueDelete(ctx, pb.QueueDeleteRequest_builder{QueueId: q.QueueId}.Build())
	return err
}

// Clear removes all objects from a Queue partition.
func (q *Queue) Clear(ctx context.Context, options *QueueClearOptions) error {
	if options == nil {
		options = &QueueClearOptions{}
	}
	if options.Partition != "" && options.All {
		return InvalidError{"options.Partition must be \"\" when clearing all partitions"}
	}
	key, err := validatePartitionKey(options.Partition)
	if err != nil {
		return err
	}
	_, err = q.client.cpClient.QueueClear(ctx, pb.QueueClearRequest_builder{
		QueueId:       q.QueueId,
		PartitionKey:  key,
		AllPartitions: options.All,
	}.Build())
	return err
}

// get is an internal helper for both Get and GetMany.
func (q *Queue) get(ctx context.Context, n int, options *QueueGetOptions) ([]any, error) {
	if options == nil {
		options = &QueueGetOptions{}
	}
	partitionKey, err := validatePartitionKey(options.Partition)
	if err != nil {
		return nil, err
	}

	startTime := time.Now()
	pollTimeout := 50 * time.Second
	if options.Timeout != nil && pollTimeout > *options.Timeout {
		pollTimeout = *options.Timeout
	}

	for {
		resp, err := q.client.cpClient.QueueGet(ctx, pb.QueueGetRequest_builder{
			QueueId:      q.QueueId,
			PartitionKey: partitionKey,
			Timeout:      float32(pollTimeout.Seconds()),
			NValues:      int32(n),
		}.Build())
		if err != nil {
			return nil, err
		}
		if len(resp.GetValues()) > 0 {
			out := make([]any, len(resp.GetValues()))
			for i, raw := range resp.GetValues() {
				v, err := pickleDeserialize(raw)
				if err != nil {
					return nil, err
				}
				out[i] = v
			}
			return out, nil
		}
		if options.Timeout != nil {
			remaining := *options.Timeout - time.Since(startTime)
			if remaining <= 0 {
				message := fmt.Sprintf("Queue %s did not return values within %s", q.QueueId, *options.Timeout)
				return nil, QueueEmptyError{message}
			}
			pollTimeout = min(pollTimeout, remaining)
		}
	}
}

// Get removes and returns one item (blocking by default).
//
// By default, this will wait until at least one item is present in the Queue.
// If `timeout` is set, returns `QueueEmptyError` if no items are available
// within that timeout in milliseconds.
func (q *Queue) Get(ctx context.Context, options *QueueGetOptions) (any, error) {
	vals, err := q.get(ctx, 1, options)
	if err != nil {
		return nil, err
	}
	return vals[0], nil // guaranteed len>=1
}

// GetMany removes up to n items.
//
// By default, this will wait until at least one item is present in the Queue.
// If `timeout` is set, returns `QueueEmptyError` if no items are available
// within that timeout in milliseconds.
func (q *Queue) GetMany(ctx context.Context, n int, options *QueueGetOptions) ([]any, error) {
	return q.get(ctx, n, options)
}

// put is an internal helper for both Put and PutMany.
func (q *Queue) put(ctx context.Context, values []any, options *QueuePutOptions) error {
	if options == nil {
		options = &QueuePutOptions{}
	}
	key, err := validatePartitionKey(options.Partition)
	if err != nil {
		return err
	}

	valuesEncoded := make([][]byte, len(values))
	for i, v := range values {
		b, err := pickleSerialize(v)
		if err != nil {
			return err
		}
		valuesEncoded[i] = b.Bytes()
	}

	deadline := time.Time{}
	if options.Timeout != nil {
		deadline = time.Now().Add(*options.Timeout)
	}

	delay := queueInitialPutBackoff
	ttl := options.PartitionTtl
	if ttl == 0 {
		ttl = queueDefaultPartitionTtl
	}

	for {
		_, err := q.client.cpClient.QueuePut(ctx, pb.QueuePutRequest_builder{
			QueueId:             q.QueueId,
			Values:              valuesEncoded,
			PartitionKey:        key,
			PartitionTtlSeconds: int32(ttl.Seconds()),
		}.Build())
		if err == nil {
			return nil // success
		}

		if status.Code(err) != codes.ResourceExhausted {
			return err
		}

		// Queue is full, retry with exponential backoff up to the deadline.
		delay = min(delay*2, 30*time.Second)
		if !deadline.IsZero() {
			remaining := time.Until(deadline)
			if remaining <= 0 {
				return QueueFullError{fmt.Sprintf("Put failed on %s", q.QueueId)}
			}
			delay = min(delay, remaining)
		}
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-time.After(delay):
		}
	}
}

// Put adds a single item to the end of the Queue.
//
// If the Queue is full, this will retry with exponential backoff until the
// provided `timeout` is reached, or indefinitely if `timeout` is not set.
// Raises `QueueFullError` if the Queue is still full after the timeout.
func (q *Queue) Put(ctx context.Context, v any, options *QueuePutOptions) error {
	return q.put(ctx, []any{v}, options)
}

// PutMany adds multiple items to the end of the Queue.
//
// If the Queue is full, this will retry with exponential backoff until the
// provided `timeout` is reached, or indefinitely if `timeout` is not set.
// Raises `QueueFullError` if the Queue is still full after the timeout.
func (q *Queue) PutMany(ctx context.Context, values []any, options *QueuePutOptions) error {
	return q.put(ctx, values, options)
}

// Len returns the number of objects in the Queue.
func (q *Queue) Len(ctx context.Context, options *QueueLenOptions) (int, error) {
	if options == nil {
		options = &QueueLenOptions{}
	}
	if options.Partition != "" && options.Total {
		return 0, InvalidError{"partition must be empty when requesting total length"}
	}
	key, err := validatePartitionKey(options.Partition)
	if err != nil {
		return 0, err
	}
	resp, err := q.client.cpClient.QueueLen(ctx, pb.QueueLenRequest_builder{
		QueueId:      q.QueueId,
		PartitionKey: key,
		Total:        options.Total,
	}.Build())
	if err != nil {
		return 0, err
	}
	return int(resp.GetLen()), nil
}

// Iterate yields items from the Queue until it is empty.
func (q *Queue) Iterate(ctx context.Context, options *QueueIterateOptions) iter.Seq2[any, error] {
	if options == nil {
		options = &QueueIterateOptions{}
	}

	itemPoll := options.ItemPollTimeout
	lastEntryID := ""
	maxPoll := 30 * time.Second

	return func(yield func(any, error) bool) {
		key, err := validatePartitionKey(options.Partition)
		if err != nil {
			yield(nil, err)
			return
		}

		fetchDeadline := time.Now().Add(itemPoll)
		for {
			pollDuration := max(0, min(maxPoll, time.Until(fetchDeadline)))
			resp, err := q.client.cpClient.QueueNextItems(ctx, pb.QueueNextItemsRequest_builder{
				QueueId:         q.QueueId,
				PartitionKey:    key,
				ItemPollTimeout: float32(pollDuration.Seconds()),
				LastEntryId:     lastEntryID,
			}.Build())
			if err != nil {
				yield(nil, err)
				return
			}
			if len(resp.GetItems()) > 0 {
				for _, item := range resp.GetItems() {
					v, err := pickleDeserialize(item.GetValue())
					if err != nil {
						yield(nil, err)
						return
					}
					if !yield(v, nil) {
						return
					}
					lastEntryID = item.GetEntryId()
				}
				fetchDeadline = time.Now().Add(itemPoll)
			} else if time.Now().After(fetchDeadline) {
				return // exit on idle
			}
		}
	}
}
